### ISSUE TITLE
Restarts ignores `val_check_interval` 

### ISSUE THREAD

**Issue by ibeltagy** [PyTorchLightning/pytorch-lightning#1182]
## 🐛 Bug

With `val_check_interval != 1`, checkpoints are saved in the middle of the epoch, but that location is not saved in the checkpoint, and after a restart, it always begins from the beginning of the last epoch. 

To reproduce: 

1. Run any training with `val_check_interval=0.2`
2. Kill training after, say, 40% of an epoch
3. Restart from the saved checkpoint
4. You will find that it is starting from 0% of that checkpoint

Solution: 

1. Save and load `batch_idx`
2. Save and load the dataloader (or just the sampler). According to this https://discuss.pytorch.org/t/how-to-save-dataloader/62813/4, it is just `torch.save(dataloader)`, but I haven't tried.

**Comment by Borda** [comment]
saving also `batch_idx` sounds reasonable for me...
any other suggestions @PyTorchLightning/core-contributors?

**Comment by ethanwharris** [comment]
This has a potential issue - we won't be able to get the shuffled data in the same order on a second pass. So even if we save the batch index, we still won't be able to do a proper resume. I suppose the best we can do is raise a warning saying that the behaviour may not be repeatable and start from the begininning of the epoch (only way to guarantee that all data is seen at least a certain number of times)  - alternatively, can we guarantee that random dataloader state is restored properly?
